src/PowerShellEditorServices/Language/FoldingReference.cs

```diff
@@ -0,0 +1,97 @@
+
+        /// <summary>
+        /// The zero-based character offset from where the folded range starts. If not defined, defaults to the length of the start line.
+        /// </summary>
+        public int StartCharacter { get; set; }
+
+        /// <summary>
+        /// The zero-based line number where the folded range ends.
+        /// </summary>
+        public int EndLine { get; set; }
+
+        /// <summary>
+        /// The zero-based character offset before the folded range ends. If not defined, defaults to the length of the end line.
+        /// </summary>
+        public int EndCharacter { get; set; }
+
+        /// <summary>
+        /// Describes the kind of the folding range such as `comment' or 'region'.
+        /// </summary>
+        public string Kind { get; set; }
+
+        /// <summary>
+        /// Builds a folding reference for start and end line and character position
+        /// </summary>
+        public FoldingReference(
+            int startLine,
+            int startCharacter,
+            int endLine,
+            int endCharacter,
+            string kind)
+        {
+            this.EndCharacter = endCharacter;
+            this.EndLine = endLine;
+            this.Kind = kind;
+            this.StartCharacter = startCharacter;
+            this.StartLine = startLine;
+        }
+
+        /// <summary>
+        /// Builds a folding reference for start and end line
+        /// </summary>
+        public FoldingReference(
+            int startLine,
+            int endLine,
+            string kind)
+        {
+            this.EndCharacter = 0;
+            this.EndLine = endLine;
+            this.Kind = kind;
+            this.StartCharacter = 0;
+            this.StartLine = startLine;
+        }
+
+        /// <summary>
+        /// A custom comparable method which can properly sort FoldingReference objects
+        /// </summary>
+        public int CompareTo(FoldingReference that) {
+            // Initially look at the start line
+            if (this.StartLine < that.StartLine) { return -1; }
+            if (this.StartLine > that.StartLine) { return 1; }
+
+            // They have the same start line so now consider the end line.
+            // The biggest line range is sorted first
+            if (this.EndLine > that.EndLine) { return -1; }
+            if (this.EndLine < that.EndLine) { return 1; }
+
+            // They have the same lines, but what about character offsets
+            if (this.StartCharacter < that.StartCharacter) { return -1; }
+            if (this.StartCharacter > that.StartCharacter) { return 1; }
+            if (this.EndCharacter < that.EndCharacter) { return -1; }
+            if (this.EndCharacter > that.EndCharacter) { return 1; }
+
+            // They're the same range, but what about kind
+            // Check for nulls
+            if ((this.Kind == null) && (that.Kind == null)) { return 0; }
+            return this.Kind.CompareTo(that.Kind);
```

Comment:
The comparison above it can be simplified to:

```csharp
if (this.Kind == that.Kind) { return 0; }
```

Now it covers both `null` and comparing an object to itself. Since these are strings they should be interned by the CLR, so referential equality will work very nicely.

Reply:
Both lines of code here can just be made into [`string.Compare(this.Kind, that.Kind)`](https://docs.microsoft.com/en-us/dotnet/api/system.string.compare?view=netframework-4.7.2#System_String_Compare_System_String_System_String_) -- that will deal with nulls and references and anything else.

Reply:
Ahh of course. I had to use `CompareTo` previously becuase I had `enumeration`.  Now that I'm using plain strings a bare `==` is fine.

Although I don't think `null == null` equates to true.

Reply:
`null == null` should be true since it's just a referential comparison. .NET doesn't do any coercive trickery on things like that, the model is just classical machine memory.

If you set `string s = "Banana";`, the true value of `s` might be `0x000f983a` or something, pointing to a struct allocated on the heap somewhere containing the string. `null` is just `0x00000000` -- `null == null` is the same test as any referential equality, just with the reference being `0`. Obviously in reality the CLR is probably slightly more complicated than this, but that's fundamentally how it works.

([`NaN == NaN`](https://dorey.github.io/JavaScript-Equality-Table/) in JS is another story though.)

Anyway, `string.Compare(s1, s2)` (`string` being the static type alias rather than a variable) will handle all of that internally, including `null`s.

Reply:
Fixed